pkg/cmd/auth/status/status.go

```diff
@@ -19,113 +19,97 @@ import (
+		)
+		activeStr := fmt.Sprintf("%v", e.Active)
+		sb.WriteString(fmt.Sprintf("  - Active account: %s\n", cs.Bold(activeStr)))
+		sb.WriteString(fmt.Sprintf("  - Git operations protocol: %s\n", cs.Bold(e.GitProtocol)))
+		sb.WriteString(fmt.Sprintf("  - Token: %s\n", cs.Bold(e.displayToken())))
+
+		if expectScopes(e.Token) {
+			sb.WriteString(fmt.Sprintf("  - Token scopes: %s\n", cs.Bold(displayScopes(e.Scopes))))
+			if err := shared.HeaderHasMinimumScopes(e.Scopes); err != nil {
+				var missingScopesError *shared.MissingScopesError
+				if errors.As(err, &missingScopesError) {
+					missingScopes := strings.Join(missingScopesError.MissingScopes, ",")
+					sb.WriteString(fmt.Sprintf("  %s Missing required token scopes: %s\n",
+						cs.WarningIcon(),
+						cs.Bold(displayScopes(missingScopes))))
+					refreshInstructions := fmt.Sprintf("gh auth refresh -h %s", e.Host)
+					sb.WriteString(fmt.Sprintf("  - To request missing scopes, run: %s\n", cs.Bold(refreshInstructions)))
+				}
+			}
+		}
 
-	if e.user != "" {
-		sb.WriteString(fmt.Sprintf("  %s Failed to log in to %s account %s (%s)\n", cs.Red("X"), e.host, cs.Bold(e.user), e.tokenSource))
-	} else {
-		sb.WriteString(fmt.Sprintf("  %s Failed to log in to %s using token (%s)\n", cs.Red("X"), e.host, e.tokenSource))
-	}
-	activeStr := fmt.Sprintf("%v", e.active)
-	sb.WriteString(fmt.Sprintf("  - Active account: %s\n", cs.Bold(activeStr)))
-	sb.WriteString(fmt.Sprintf("  - The token in %s is invalid.\n", e.tokenSource))
-	if e.tokenIsWriteable {
-		loginInstructions := fmt.Sprintf("gh auth login -h %s", e.host)
-		logoutInstructions := fmt.Sprintf("gh auth logout -h %s -u %s", e.host, e.user)
-		sb.WriteString(fmt.Sprintf("  - To re-authenticate, run: %s\n", cs.Bold(loginInstructions)))
-		sb.WriteString(fmt.Sprintf("  - To forget about this account, run: %s\n", cs.Bold(logoutInstructions)))
-	}
-
-	return sb.String()
-}
-
-type timeoutErrorEntry struct {
-	active      bool
-	host        string
-	user        string
-	tokenSource string
-}
-
-func (e timeoutErrorEntry) String(cs *iostreams.ColorScheme) string {
-	var sb strings.Builder
+	case authStateTimeout:
+		if e.Login != "" {
+			sb.WriteString(fmt.Sprintf("  %s Timeout trying to log in to %s account %s (%s)\n", cs.Red("X"), e.Host, cs.Bold(e.Login), e.TokenSource))
+		} else {
+			sb.WriteString(fmt.Sprintf("  %s Timeout trying to log in to %s using token (%s)\n", cs.Red("X"), e.Host, e.TokenSource))
+		}
+		activeStr := fmt.Sprintf("%v", e.Active)
+		sb.WriteString(fmt.Sprintf("  - Active account: %s\n", cs.Bold(activeStr)))
+
+	case authStateError:
+		if e.Login != "" {
+			sb.WriteString(fmt.Sprintf("  %s Failed to log in to %s account %s (%s)\n", cs.Red("X"), e.Host, cs.Bold(e.Login), e.TokenSource))
+		} else {
+			sb.WriteString(fmt.Sprintf("  %s Failed to log in to %s using token (%s)\n", cs.Red("X"), e.Host, e.TokenSource))
+		}
```

Comment:
I tried `gh auth status` with and without `--json`, and noticed the output is not the same (i.e. `error` field was empty for invalid entries). The reason is `authEntry.Error` is never assigned in `buildEntry`, and rather we print the error here in display stage.

So, let's move these error messages to `buildEntry` and populate `authEntry.Error` with them, and then here we can just print the `authEntry.Error` value.

Reply:
good catch! I updated the test to cover the JSON output of the error field, plus added a missing test on the token auth. 
Also, I refactored the String function, which was hard to read, without the switch/case.

I chose to keep a minimal error in the JSON, and append the formatted details (login, host and token source) only for the regular output, as all of these information are already available in the JSON. Wdyt?